cli/util/uploader.go

```diff
@@ -0,0 +1,432 @@
+	indexFileSize := ctx.Uint("index-file-size")
+	fmt.Fprintln(ctx.App.Writer, "Updating index files...")
+
+	prm := client.PrmObjectSearch{}
+	filters := object.NewSearchFilters()
+	filters.AddFilter(attributeKey, fmt.Sprintf("%d", 0), object.MatchNumGE)
+	filters.AddFilter("size", fmt.Sprintf("%d", indexFileSize), object.MatchStringEqual)
+	prm.SetFilters(filters)
+
+	objectIDs, err := neofs.ObjectSearch(ctx.Context, &p, account.PrivateKey(), containerID.String(), prm)
+	if err != nil {
+		return fmt.Errorf("search of index files failed: %w", err)
+	}
+
+	existingIndexCount := uint(len(objectIDs))
+	expectedIndexCount := currentHeight / indexFileSize
+
+	if existingIndexCount >= expectedIndexCount {
+		fmt.Fprintf(ctx.App.Writer, "Index files are up to date. Existing: %d, expected: %d\n", existingIndexCount, expectedIndexCount)
+		return nil
+	}
+
+	var (
+		processedCounter atomic.Int32
+		errCh            = make(chan error, 1)
+		buffer           = make([]byte, indexFileSize*oidSize)
+		oidCh            = make(chan oid.ID, indexFileSize)
+	)
+
+	for range maxParallelSearches {
+		go func() {
+			for currOid := range oidCh {
+				obj, err := p.ObjectHead(context.Background(), containerID, currOid, signer, client.PrmObjectHead{})
+				if err != nil {
+					select {
+					case errCh <- fmt.Errorf("failed to get header for OID %s: %w", currOid.String(), err):
+					default:
+					}
+					return
+				}
+
+				blockIndex, err := getBlockIndex(*obj, blockAttributeKey)
+				if err != nil {
+					select {
+					case errCh <- fmt.Errorf("failed to process header for OID %s: %w", currOid.String(), err):
+					default:
+					}
+					return
+				}
+
+				index := uint(blockIndex) / indexFileSize
+				offset := (uint(blockIndex) - index*indexFileSize) * oidSize
+				res := make([]byte, oidSize)
+				currOid.Encode(res)
+				copy(buffer[offset:], res)
+				processedCounter.Add(1)
+			}
+		}()
+	}
+
+	for i := existingIndexCount; i < expectedIndexCount; i++ {
+		startIndex := i * indexFileSize
+		endIndex := startIndex + indexFileSize
+
+		for j := int(startIndex); j < int(endIndex); j += searchBatchSize {
+			remaining := int(endIndex) - j
+			end := j + min(searchBatchSize, remaining)
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
+			filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			objectIDs, err = neofs.ObjectSearch(ctx.Context, &p, account.PrivateKey(), containerID.String(), prm)
+			if err != nil {
+				return fmt.Errorf("no OIDs found for index files: %w", err)
+			}
+
+			for _, currOid := range objectIDs {
+				oidCh <- currOid
+			}
+		}
+
+		for processedCounter.Load() < int32(indexFileSize) {
+			select {
+			case err := <-errCh:
+				return err
+			default:
+				time.Sleep(100 * time.Millisecond)
+			}
+		}
```

Comment:
I see, `processedCounter` doesn't allow us to wait properly. Awaiting mechanism should be improved a bit, because it's always not good to `Sleep`. Let's replace `processedCounter` with a single `chan struct{}`; every processor routine should send `struct{}` to this channel, and then we'll have something like this:
```
var oidFetcherToProcessor chan struct{}

...		
		
		var completed int
	waitLoop:
		for {
			select {
			case err := <-errCh:
				return err
			case <-oidFetcherToProcessor:
				completed++
				if completed == int(indexFileSize) {
					break waitLoop
				}
			}
		}
```

Reply:
And then you have to send the following part of code to a separate routine in order to avoid possible deadlock on error sending to errCh:
```
go func(){for j := int(startIndex); j < int(endIndex); j += searchBatchSize {...}}()
```